### PR TITLE
fix: [CDS-78547]: Added delegate selector for container step from connector and step

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
Added the merged delegate selectors from step and connector for container step. 
Tests Done:
1. Verified the merged delegate selectors coming from both connector and step.
2. Verified the delegate selectors coming from connector alone.
3. Verified the delegate selectors coming from step alone.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52598)

### CODE REVIEW

**Comment by jennyjames21** [comment]
Question: Does change follow the hierarchy mentioned in the document? https://developer.harness.io/docs/platform/delegates/manage-delegates/select-delegates-with-selectors/

**Comment by sourabh-awashti** [comment]
@jennyjames21 yes i will add the selectors from step, stage and pipeline  as well

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52598)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Review by sahilhindwani-harness — DISMISSED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52598)* status: :shipit: complete! ✓

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger ut1, ut2, ut6, ut8, ut9

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/execution/AbstractContainerStep.java [hunk 1]
```diff
@@ -100,6 +115,11 @@ public AsyncExecutableResponse executeAsyncAfterRbac(
         .build();
   }
 
+  private String getConnectorRef(ContainerStepSpec containerStepInfo) {
```
Comment (yogesh-chauhan) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/execution/AbstractContainerStep.java:118]: nit: private methods at the end of the class for better readability

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/utils/ContainerSpecUtils.java [hunk 2]
```diff
@@ -33,4 +40,29 @@ public List<TaskSelector> getDelegateSelectors(ContainerStepSpec containerStepSp
     }
     return List.of();
   }
+
+  @NotNull
+  private List<TaskSelector> getConnectorDelegateSelectors(ConnectorDetails k8sConnector) {
+    if (k8sConnector != null) {
+      Set<String> delegateSelectorSet =
+          ((KubernetesClusterConfigDTO) k8sConnector.getConnectorConfig()).getDelegateSelectors();
```
Comment (yogesh-chauhan) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/utils/ContainerSpecUtils.java:48]: I think we should do a type check here that is a k8s connector

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/execution/AbstractContainerStep.java (hunk 1 above)
Comment (sourabh-awashti) [reply pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/execution/AbstractContainerStep.java:118]: done

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/utils/ContainerSpecUtils.java (hunk 2 above)
Comment (sourabh-awashti) [reply pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/utils/ContainerSpecUtils.java:48]: done

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/execution/AbstractContainerStep.java [hunk 3]
```diff
@@ -153,4 +168,9 @@ private void abortTasks(List<String> allCallbackIds, String callbackId) {
                 .errorMessage("Delegate is not able to connect to created build farm")
                 .build()));
   }
+
+  private String getConnectorRef(ContainerStepSpec containerStepInfo) {
+    ContainerK8sInfra containerK8sInfra = (ContainerK8sInfra) containerStepInfo.getInfrastructure();
```
Comment (yogesh-chauhan) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/container/execution/AbstractContainerStep.java:173]: we should do a type check here

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 9 of 11 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52598)* status: 11 file(s) to review. 1. unresolved discussions.

**Comment by sourabh-awashti** [comment]
trigger review

**Pull request merged**